File: commands/stats/milestones.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
import re
from database import db

class Milestones(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='milestones', description='Show clan milestones and achievements')
    async def milestones(self, interaction: discord.Interaction, clan: str):
        try:
            await interaction.response.defer()
            
            clan_upper = clan.upper().strip()
            query = {'clan_name': {'$regex': f'^{re.escape(clan_upper)}$', '$options': 'i'}}
            
            pipeline = [
                {'$match': query},
                {'$group': {
                    '_id': None,
                    'total': {'$sum': 1},
                    'contests': {'$sum': {'$cond': ['$contest', 1, 0]}},
                    'points': {'$sum': {'$cond': ['$contest', {'$multiply': ['$player_count', 2]}, '$player_count']}}
                }}
            ]
            
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            if not result:
                await interaction.followup.send(embed=discord.Embed(
                    title=f"[{clan_upper}] Milestones", description="No data found.", color=0xff0000
                ))
                return
            
            stats = result[0]
            total_wins = stats.get('total', 0)
            contest_wins = stats.get('contests', 0)
            total_points = stats.get('points', 0)
            
            win_milestones = [10, 25, 50, 100, 250, 500, 1000]
            contest_milestones = [5, 10, 25, 50, 100, 250]
            point_milestones = [100, 500, 1000, 2500, 5000, 10000]
            
            highest_win = max([m for m in win_milestones if total_wins >= m], default=0)
            highest_contest = max([m for m in contest_milestones if contest_wins >= m], default=0)
            highest_point = max([m for m in point_milestones if total_points >= m], default=0)
            
            next_win = next((m for m in win_milestones if total_wins < m), None)
            next_contest = next((m for m in contest_milestones if contest_wins < m), None)
            next_point = next((m for m in point_milestones if total_points < m), None)
            
            embed = discord.Embed(
                title=f"[{clan_upper}] Milestones",
                description=f"{total_wins} wins | {contest_wins} contests | {total_points} points",
                color=0xffd700
            )
            
            achievements = []
            if highest_win > 0:
                achievements.append(f"🏆 {highest_win} Wins")
            if highest_contest > 0:
                achievements.append(f"🎯 {highest_contest} Contests")
            if highest_point > 0:
                achievements.append(f"💰 {highest_point} Points")
            
            if achievements:
                embed.add_field(name="🏅 Current Badges", value="\n".join(achievements), inline=False)
            
            next_goals = []
            if next_win:
                next_goals.append(f"🏆 {next_win} Wins ({next_win - total_wins} to go)")
            if next_contest:
                next_goals.append(f"🎯 {next_contest} Contests ({next_contest - contest_wins} to go)")
            if next_point:
                next_goals.append(f"💰 {next_point} Points ({next_point - total_points} to go)")
            
            if next_goals:
                embed.add_field(name="🎯 Next Goals", value="\n".join(next_goals), inline=False)
            else:
                embed.add_field(name="🎯 Next Goals", value="All milestones unlocked!", inline=False)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in milestones: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to fetch milestones.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(Milestones(bot))

File: commands/stats/milestones.js
// https://github.com/viktorexe/terristats-discord-bot
// /milestones - achievement badges a clan has unlocked and the next goals.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { exactMatch, gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('milestones')
    .setDescription('Show clan milestones and achievements')
    .addStringOption((o) =>
      o.setName('clan').setDescription('Clan name').setRequired(true),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const clan = interaction.options.getString('clan');
      const clanUpper = clan.toUpperCase().trim();
      const query = { clan_name: exactMatch(clanUpper) };

      const pipeline = [
        { $match: query },
        {
          $group: {
            _id: null,
            total: { $sum: 1 },
            contests: { $sum: { $cond: ['$contest', 1, 0] } },
            points: {
              $sum: {
                $cond: [
                  '$contest',
                  { $multiply: ['$player_count', 2] },
                  '$player_count',
                ],
              },
            },
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      if (!result.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`[${clanUpper}] Milestones`)
              .setDescription('No data found.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const stats = result[0];
      const totalWins = stats.total || 0;
      const contestWins = stats.contests || 0;
      const totalPoints = stats.points || 0;

      const winMilestones = [10, 25, 50, 100, 250, 500, 1000];
      const contestMilestones = [5, 10, 25, 50, 100, 250];
      const pointMilestones = [100, 500, 1000, 2500, 5000, 10000];

      const highestOf = (list, value) =>
        list.filter((m) => value >= m).reduce((a, b) => Math.max(a, b), 0);
      const nextOf = (list, value) => list.find((m) => value < m) ?? null;

      const highestWin = highestOf(winMilestones, totalWins);
      const highestContest = highestOf(contestMilestones, contestWins);
      const highestPoint = highestOf(pointMilestones, totalPoints);

      const nextWin = nextOf(winMilestones, totalWins);
      const nextContest = nextOf(contestMilestones, contestWins);
      const nextPoint = nextOf(pointMilestones, totalPoints);

      const embed = new EmbedBuilder()
        .setTitle(`[${clanUpper}] Milestones`)
        .setDescription(`${totalWins} wins | ${contestWins} contests | ${totalPoints} points`)
        .setColor(0xffd700);

      const achievements = [];
      if (highestWin > 0) achievements.push(`🏆 ${highestWin} Wins`);
      if (highestContest > 0) achievements.push(`🎯 ${highestContest} Contests`);
      if (highestPoint > 0) achievements.push(`💰 ${highestPoint} Points`);
      if (achievements.length) {
        embed.addFields({ name: '🏅 Current Badges', value: achievements.join('\n'), inline: false });
      }

      const nextGoals = [];
      if (nextWin) nextGoals.push(`🏆 ${nextWin} Wins (${nextWin - totalWins} to go)`);
      if (nextContest) nextGoals.push(`🎯 ${nextContest} Contests (${nextContest - contestWins} to go)`);
      if (nextPoint) nextGoals.push(`💰 ${nextPoint} Points (${nextPoint - totalPoints} to go)`);
      embed.addFields({
        name: '🎯 Next Goals',
        value: nextGoals.length ? nextGoals.join('\n') : 'All milestones unlocked!',
        inline: false,
      });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in milestones: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to fetch milestones.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
